File: smoke_tests/php/cryptographic_failures.php
<?php
// Cryptographic Failures in PHP

// Test 1: ECB mode
function encrypt_ecb($data) {
    $key = 'sixteen_byte_key';
    // VULNERABLE: ECB mode reveals patterns
    return openssl_encrypt($data, 'AES-128-ECB', $key);
}

// Test 2: Hardcoded encryption key
function encrypt_hardcoded($data) {
    // VULNERABLE: Hardcoded key
    $key = '0123456789abcdef';
    $iv = openssl_random_pseudo_bytes(16);
    return openssl_encrypt($data, 'AES-128-CBC', $key, 0, $iv);
}

// Test 3: Static IV
function encrypt_static_iv($data) {
    $key = random_bytes(16);
    // VULNERABLE: Static IV
    $iv = str_repeat("\x00", 16);
    return openssl_encrypt($data, 'AES-128-CBC', $key, 0, $iv);
}

// Test 4: DES usage
function encrypt_des($data) {
    $key = 'password';
    $iv = openssl_random_pseudo_bytes(8);
    // VULNERABLE: DES is weak
    return openssl_encrypt($data, 'DES-CBC', $key, 0, $iv);
}

// Test 5: MD5 for integrity
function hash_md5($data) {
    // VULNERABLE: MD5 is broken
    return md5($data);
}

// Test 6: SHA1 for security
function hash_sha1($data) {
    // VULNERABLE: SHA1 is deprecated
    return sha1($data);
}

// Test 7: Weak password hash
function hash_password_weak($password) {
    // VULNERABLE: MD5 for password
    return md5($password);
}

// Test 8: Unsalted password hash
function hash_password_unsalted($password) {
    // VULNERABLE: No salt
    return hash('sha256', $password);
}

// Test 9: Insufficient PBKDF2 iterations
function derive_key($password) {
    $salt = random_bytes(16);
    // VULNERABLE: Only 1000 iterations
    return hash_pbkdf2('sha256', $password, $salt, 1000, 32, true);
}

// Test 10: Static salt
function hash_with_static_salt($password) {
    // VULNERABLE: Static salt
    $salt = 'constant_salt_value';
    return hash('sha256', $salt . $password);
}

// Test 11: rand() for crypto
function generate_key_weak() {
    // VULNERABLE: rand() is not crypto secure
    $key = '';
    for ($i = 0; $i < 16; $i++) {
        $key .= chr(rand(0, 255));
    }
    return $key;
}

// Test 12: Password as key directly
function encrypt_with_password($data, $password) {
    // VULNERABLE: Password should go through KDF
    $key = substr($password . str_repeat("\x00", 16), 0, 16);
    $iv = random_bytes(16);
    return openssl_encrypt($data, 'AES-128-CBC', $key, 0, $iv);
}

// Test 13: mcrypt (deprecated)
function encrypt_mcrypt($data) {
    // VULNERABLE: mcrypt is deprecated and insecure
    // $key = 'secretkey';
    // $iv = mcrypt_create_iv(16, MCRYPT_DEV_RANDOM);
    // return mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $key, $data, MCRYPT_MODE_CBC, $iv);
}

// Test 14: Insecure key storage
function get_encryption_key() {
    // VULNERABLE: Key stored in code
    return 'my_secret_encryption_key_12345';
}

// Test 15: RC4 usage
function encrypt_rc4($data) {
    $key = random_bytes(16);
    // VULNERABLE: RC4 is broken
    return openssl_encrypt($data, 'RC4', $key);
}
?>

File: smoke_tests/ruby/cryptographic_failures.rb
# Cryptographic Failures in Ruby
require 'openssl'
require 'digest'
require 'base64'

class CryptoController < ApplicationController
  # Test 1: ECB mode
  def encrypt_ecb
    data = params[:data]
    cipher = OpenSSL::Cipher.new('AES-128-ECB')
    cipher.encrypt
    # VULNERABLE: ECB mode reveals patterns
    cipher.key = 'sixteen byte key'
    encrypted = cipher.update(data) + cipher.final
    render plain: Base64.encode64(encrypted)
  end

  # Test 2: Hardcoded encryption key
  def encrypt_with_key
    data = params[:data]
    cipher = OpenSSL::Cipher.new('AES-128-CBC')
    cipher.encrypt
    # VULNERABLE: Hardcoded key
    cipher.key = '0123456789abcdef'
    cipher.iv = cipher.random_iv
    encrypted = cipher.update(data) + cipher.final
    render plain: Base64.encode64(encrypted)
  end

  # Test 3: Static IV
  def encrypt_static_iv
    data = params[:data]
    cipher = OpenSSL::Cipher.new('AES-128-CBC')
    cipher.encrypt
    cipher.key = OpenSSL::Random.random_bytes(16)
    # VULNERABLE: Static IV
    cipher.iv = "\x00" * 16
    encrypted = cipher.update(data) + cipher.final
    render plain: Base64.encode64(encrypted)
  end

  # Test 4: DES usage
  def encrypt_des
    data = params[:data]
    # VULNERABLE: DES is weak
    cipher = OpenSSL::Cipher.new('DES-CBC')
    cipher.encrypt
    cipher.key = 'password'
    cipher.iv = cipher.random_iv
    encrypted = cipher.update(data) + cipher.final
    render plain: Base64.encode64(encrypted)
  end

  # Test 5: MD5 for integrity
  def hash_md5
    data = params[:data]
    # VULNERABLE: MD5 is broken
    hash = Digest::MD5.hexdigest(data)
    render plain: hash
  end

  # Test 6: SHA1 for security
  def hash_sha1
    data = params[:data]
    # VULNERABLE: SHA1 is deprecated
    hash = Digest::SHA1.hexdigest(data)
    render plain: hash
  end

  # Test 7: Small RSA key
  def generate_rsa
    # VULNERABLE: 1024-bit RSA is too weak
    key = OpenSSL::PKey::RSA.new(1024)
    render plain: key.public_key.to_pem
  end

  # Test 8: Weak PBKDF2 iterations
  def derive_key
    password = params[:password]
    salt = OpenSSL::Random.random_bytes(16)
    # VULNERABLE: Only 1000 iterations
    key = OpenSSL::PKCS5.pbkdf2_hmac(password, salt, 1000, 32, 'sha256')
    render plain: Base64.encode64(key)
  end

  # Test 9: Static salt
  def hash_with_salt
    password = params[:password]
    # VULNERABLE: Static salt
    salt = 'constant_salt_value'
    hash = OpenSSL::PKCS5.pbkdf2_hmac(password, salt, 100_000, 32, 'sha256')
    render plain: Base64.encode64(hash)
  end

  # Test 10: No authenticated encryption
  def encrypt_no_auth
    data = params[:data]
    cipher = OpenSSL::Cipher.new('AES-128-CBC')
    cipher.encrypt
    cipher.key = OpenSSL::Random.random_bytes(16)
    cipher.iv = OpenSSL::Random.random_bytes(16)
    # VULNERABLE: No authentication (padding oracle possible)
    encrypted = cipher.update(data) + cipher.final
    render plain: Base64.encode64(encrypted)
  end

  # Test 11: rand for crypto
  def generate_weak_key
    # VULNERABLE: rand is not cryptographically secure
    key = (0...16).map { rand(256).chr }.join
    render plain: Base64.encode64(key)
  end

  # Test 12: Password as key directly
  def encrypt_with_password
    password = params[:password]
    data = params[:data]
    cipher = OpenSSL::Cipher.new('AES-128-CBC')
    cipher.encrypt
    # VULNERABLE: Password should go through KDF
    cipher.key = password.ljust(16, "\x00")[0..15]
    cipher.iv = OpenSSL::Random.random_bytes(16)
    encrypted = cipher.update(data) + cipher.final
    render plain: Base64.encode64(encrypted)
  end
end
